models/git_blame.go

```diff
@@ -0,0 +1,101 @@
+// Copyright 2019 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package models
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"regexp"
+
+	"code.gitea.io/git"
+	"code.gitea.io/gitea/modules/process"
+)
+
+// BlameFile represents while git blame output
+type BlameFile struct {
+	Parts []BlamePart
+}
+
+// BlamePart represents block of blame - continuous lines with one sha
+type BlamePart struct {
+	Sha   string
+	Lines []string
+}
+
+var blameLineRegex = regexp.MustCompile(`^([a-z0-9]*)\s*(\S*)\s*(\d*)\) (.*)`)
+
+func parseBlameOutput(reader io.Reader) (*BlameFile, error) {
```

Comment:
I think refactor this so you return say a `BlameReader` which similar to say `bufio.Reader` but with a function `ReadBlameLine` which returns each blame line or part in turn.

Remember files in the git repository can be arbitrarily large. (I'm sure there's other places that load the whole file in gitea but we should stop doing it if we can.)